src/org/opensolaris/opengrok/util/OptionParser.java

```diff
@@ -0,0 +1,701 @@
+            if(isOption ) {
+                value = null;
+            }
+        } else if (isOption) {
+            value = "";
+        }
+        return value;
+    }
+    
+    private String getOption(String arg) {
+        String option = null;
+        
+        if (arg.startsWith("-")) {
+            if (arg.startsWith("--")) {
+                option = arg;                 // long name option (--longOption)
+            } else if (arg.length() > 2) {
+                option = arg.substring(0,2);  // short name option (-xValue)
+            } else {
+                option = arg;                 // short name option (-x)
+            }
+        }
+        return option;
+    }
+
+    // Allows user to enter initial substring of a long option name.
+    private boolean unknown(String option, int index) throws ParseException {
+        boolean found = options.containsKey(option);
+        List<String> candidates = new ArrayList<String>();
+        
+        if (!found) {
+            // Now check to see if initial substring was entered.
+            for (String key: options.keySet()) {
+                if (key.startsWith(option)) {
+                    candidates.add(key);
+                }
+            }
+            if (candidates.size() == 1 ) {
+                found = true;
+            } else if (candidates.size() > 1) {
+                throw new ParseException(
+                    "Ambiguous option " + option + " matches " + candidates, index);
+            }
+        }
+        return !found;
+    }
+        
+    /**
+     * Parse given set of arguments and activate handlers
+     * 
+     * This code parses the given set of parameters looking for a described
+     * set of options and activates the code segments associated with the
+     * option.
+     * 
+     * Parsing is discontinued when a lone "--" is encountered in the list of
+     * arguments. If this is a normal non-scan parser, unrecognized options
+     * will cause a parse exception. If this is a scan parser, unrecognized
+     * options are ignored.
+     * 
+     * @param args
+     * @return non-option parameters, or all arguments after "--" encountered.
+     * @throws ParseException 
+     */
+
+    public String[] parse(String[] args) throws ParseException {
+        int ii = 0;
+        int optind = -1;
+        String option;
+        while (ii < args.length) {
+            option = getOption(args[ii]);
+            
+            // When scanning for specific options...
+            if (scanning) {
+                if (option == null || unknown(option,ii)) {
+                    optind = ++ii;  // skip over everything else
+                    continue;
+                }
+            }
+
+            if (!option.startsWith("-")) {
+                break;
+            } else {
+                if (option.equals("--")) {
+                    optind = ii + 1;
+                    break;
+                }
+
+                if (unknown(option, ii)) {
```

Comment:
If `unknown()` already did the job and found partial match, it would be wasteful to repeat this work. Also, I don't see how partial matching could actually work - if `unknown()` returns `false` then the partial option would get matched in `options.get()` which cannot succeed.

Reply:
I see your point. I thought my original test covered this, but it doesn't. So I need to write another test. My cure will be to change 'unknown' into 'candidate' which will return the actual option name found, or null when not found.